azure-iot-device/azure/iot/device/common/http_transport.py

```diff
@@ -0,0 +1,193 @@
+                self._x509_cert.key_file,
+                self._x509_cert.pass_phrase,
+            )
+
+        return ssl_context
+
+    # def _format_headers(self, headers):
+    #     # TODO: Right now I'm only doing sas token because I'm lazy and limited
+    #     formatted_headers = {}
+    #     formatted_headers["Authorization"] = headers["sas"]
+    #     return formatted_headers
+
+    """
+    Some backstory on this.
+
+    1) During discussion the concept of multiple simultaneous requests being made was brought up.
+    Let's say you are a person and you want to make a request a bunch of times for some reason.
+    We could do two things down here. First option, we could have a single HTTPSConnection established
+    during instantation of the transport, in which case the HTTPSConnection would be shared among all
+    requests! Cool? Maybe. I don't know precisely, but I'm pretty sure that if the conneciton is making a
+    request, and then another thread comes along and using the same connection makes a request... well I just
+    don't really know what's going to happen. So the second (and currently implemented) option for this is to
+    create a new HTTPSConnection object every time we make a request, so that each request has it's own "connection".
+    What does this mean? Not sure, but it seem safer.
+
+    2) Initially I atomized each http_client call. We would create the client in one call, we would connect in another,
+    disconnect in yet another. Seemed fine... ish. But then Carter was like hey we should make all the http_client specific
+    stuff down here and only expose a relevant surface. What's a relevant surface for HTTP? Probably just request. So that's
+    what I'm exposing. Everything, the insantiation of the HTTPSConnection, the sending of the request, the getting of the response,
+    all happens in this one call, and that's better. Probably.
+    """
+
+    @pipeline_thread.invoke_on_http_thread_nowait
+    # TODO: This star syntax is incompatible with Python 2, change it so that the callback is in front of the optional params.
+    def request(self, method, hostname, path, callback, body=None, headers={}, query_params=None):
+        # Sends a complete request to the server
+        logger.info("sending https request")
+        try:
+            logger.debug("creating to https connection")
+            connection = http_client.HTTPSConnection(
+                self._hostname, context=self._create_ssl_context()
+            )
+            logger.debug("connecting to host tcp socket")
+            connection.connect()
+            logger.debug("connection succeeded")
+            # formatted_headers = self._format_headers(headers)
+            url = "https://{hostname}/{path}?{query_params}".format(
+                hostname=hostname, path=path, query_params=query_params
+            )
+            connection.request(method, url, body=body.encode("utf-8"), headers=headers)
```

Comment:
>encode("utf-8") [](start = 54, length = 15)

I'd like encoding to happen in a consistent place regardless of which pipeline/transport we're using. Not to say the way it's currently being done for the primary hub/MQTT pipeline is necessarily more correct than this, but it shouldn't be inconsistent. In MQTT we do this well before it gets down to the transport.

Reply:
ok, good suggestion, I'll move the encoding further up.

---
In reply to: [351475147](https://github.com/Azure/azure-iot-sdk-python/pull/403#discussion_r351475147) [](ancestors = 351475147)